storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/KvStoreDatabase.java

```diff
@@ -813,49 +793,23 @@ public void removeBlobSidecars(final UInt64 slot) {
   }
 
   @Override
-  public void removeBlobsSidecar(final SlotAndBlockRoot slotAndBlockRoot) {
-    try (final FinalizedUpdater updater = finalizedUpdater()) {
-      updater.removeBlobsSidecar(slotAndBlockRoot);
-      updater.confirmBlobsSidecar(slotAndBlockRoot);
-      updater.commit();
-    }
-  }
-
-  @Override
-  public boolean pruneOldestBlobsSidecar(final UInt64 lastSlotToPrune, final int pruneLimit) {
-    try (final Stream<SlotAndBlockRoot> prunableBlobsKeys =
-            streamBlobsSidecarKeys(UInt64.ZERO, lastSlotToPrune);
+  public boolean pruneOldestBlobSidecars(final UInt64 lastSlotToPrune, final int pruneLimit) {
+    try (final Stream<SlotAndBlockRootAndBlobIndex> prunableBlobKeys =
+            streamBlobSidecarKeys(UInt64.ZERO, lastSlotToPrune);
         final FinalizedUpdater updater = finalizedUpdater()) {
-      final long pruned =
-          prunableBlobsKeys
-              .limit(pruneLimit)
-              .peek(
-                  slotAndBlockRoot -> {
-                    updater.removeBlobsSidecar(slotAndBlockRoot);
-                    updater.confirmBlobsSidecar(slotAndBlockRoot);
-                  })
-              .count();
-      updater.commit();
-
-      return pruned == pruneLimit;
-    }
-  }
-
-  @Override
-  public boolean pruneOldestUnconfirmedBlobsSidecars(
-      final UInt64 lastSlotToPrune, final int pruneLimit) {
-    try (final Stream<SlotAndBlockRoot> prunableUnconfirmed =
-            streamUnconfirmedBlobsSidecars(UInt64.ZERO, lastSlotToPrune);
-        final FinalizedUpdater updater = finalizedUpdater()) {
-      final long pruned =
-          prunableUnconfirmed
-              .limit(pruneLimit)
-              .peek(
-                  slotAndBlockRoot -> {
-                    updater.removeBlobsSidecar(slotAndBlockRoot);
-                    updater.confirmBlobsSidecar(slotAndBlockRoot);
-                  })
-              .count();
+      int remaining = pruneLimit;
+      int pruned = 0;
+      for (final Iterator<SlotAndBlockRootAndBlobIndex> it = prunableBlobKeys.iterator();
+          it.hasNext(); ) {
+        --remaining;
+        final boolean finished = remaining < 0;
+        final SlotAndBlockRootAndBlobIndex key = it.next();
+        if (finished && (key.isNoBlobsKey() || key.getBlobIndex().equals(ZERO))) {
```

Comment:
So, IIUC, we want to continue pruning blobs for the current slot so we don't partially delete blobs for a given slot.
Correct?

if yes it means that `pruned` can go over `pruneLimit` and the return statement should be

`return pruned >= pruneLimit`

Reply:
Nice catch